tests/unit/nupic/encoders/category_test.py

```diff
@@ -36,189 +40,226 @@ class CategoryEncoderTest(unittest.TestCase):
 
 
   def testCategoryEncoder(self):
-      verbosity = 0
-
-      print "Testing CategoryEncoder...",
-      categories = ["ES", "GB", "US"]
-
+    verbosity = 0
+
+    print "Testing CategoryEncoder...",
+    categories = ["ES", "GB", "US"]
+
+    # forced: is not recommended, but is used here for readability. see scalar.py
+    e = CategoryEncoder(w=3, categoryList=categories, forced=True)
+    output = e.encode("US")
+    self.assertTrue((output == numpy.array([0,0,0,0,0,0,0,0,0,1,1,1], dtype=defaultDtype)).all())
+
+    # Test reverse lookup
+    decoded = e.decode(output)
+    (fieldsDict, fieldNames) = decoded
+    self.assertEqual(len(fieldsDict), 1)
+    (ranges, desc) = fieldsDict.values()[0]
+    self.assertTrue(len(ranges) == 1 and numpy.array_equal(ranges[0], [3,3]))
+    print "decodedToStr of", ranges, "=>", e.decodedToStr(decoded)
+
+    # Test topdown compute
+    for v in categories:
+      output = e.encode(v)
+      topDown = e.topDownCompute(output)
+      self.assertEqual(topDown.value, v)
+      self.assertEqual(topDown.scalar, e.getScalars(v)[0])
+
+      bucketIndices = e.getBucketIndices(v)
+      print "bucket index =>", bucketIndices[0]
+      topDown = e.getBucketInfo(bucketIndices)[0]
+      self.assertEqual(topDown.value, v)
+      self.assertEqual(topDown.scalar, e.getScalars(v)[0])
+      self.assertTrue((topDown.encoding == output).all())
+      self.assertEqual(topDown.value, e.getBucketValues()[bucketIndices[0]])
+
+
+
+    # ---------------------
+    # unknown category
+    output = e.encode("NA")
+    self.assertTrue((output == numpy.array([1,1,1,0,0,0,0,0,0,0,0,0], dtype=defaultDtype)).all())
+
+    # Test reverse lookup
+    decoded = e.decode(output)
+    (fieldsDict, fieldNames) = decoded
+    self.assertEqual(len(fieldsDict), 1)
+    (ranges, desc) = fieldsDict.values()[0]
+    self.assertTrue(len(ranges) == 1 and numpy.array_equal(ranges[0], [0,0]))
+    print "decodedToStr of", ranges, "=>", e.decodedToStr(decoded)
+
+    # Test topdown compute
+    topDown = e.topDownCompute(output)
+    self.assertEqual(topDown.value, "<UNKNOWN>")
+    self.assertEqual(topDown.scalar, 0)
+
+
+    # --------------------------------
+    # ES
+    output = e.encode("ES")
+    self.assertTrue((output == numpy.array([0,0,0,1,1,1,0,0,0,0,0,0], dtype=defaultDtype)).all())
+
+    # MISSING VALUE
+    outputForMissing = e.encode(SENTINEL_VALUE_FOR_MISSING_DATA)
+    self.assertEqual(sum(outputForMissing), 0)
+
+    # Test reverse lookup
+    decoded = e.decode(output)
+    (fieldsDict, fieldNames) = decoded
+    self.assertEqual(len(fieldsDict), 1)
+    (ranges, desc) = fieldsDict.values()[0]
+    self.assertTrue(len(ranges) == 1 and numpy.array_equal(ranges[0], [1,1]))
```

Comment:
I'd separate this into an `assertEqual(len(ranges), 1)` and a `self.assertTrue(numpy.array_equal(...`